src/coprocessor/dag/expr/builtin_math.rs

```diff
@@ -209,6 +209,67 @@ impl ScalarFunc {
         self.children[0].eval_int(ctx, row)
     }
 
+    #[inline]
+    pub fn round_dec<'a, 'b: 'a>(
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, Decimal>>> {
+        let d = try_opt!(self.children[0].eval_decimal(ctx, row));
+        let result: Result<Decimal> = d
+            .into_owned()
+            .round(DEFAULT_FSP, RoundMode::HalfEven)
+            .into();
+        result.map(|t| Some(Cow::Owned(t)))
+    }
+
+    #[inline]
+    pub fn round_real(&self, ctx: &mut EvalContext, row: &[Datum]) -> Result<Option<f64>> {
+        let n = try_opt!(self.children[0].eval_real(ctx, row));
+        Ok(Some(n.round()))
+    }
+
+    #[inline]
+    pub fn round_with_frac_int(&self, ctx: &mut EvalContext, row: &[Datum]) -> Result<Option<i64>> {
+        let x = try_opt!(self.children[0].eval_int(ctx, row));
+        let d = try_opt!(self.children[1].eval_int(ctx, row));
+
+        if d >= 0 {
+            Ok(Some(x))
+        } else {
+            let base = 10.0_f64.powi(-d as i32);
+            let frac = x as f64 / base;
+            Ok(Some((frac.round() * base) as i64))
+        }
+    }
+
+    #[inline]
+    pub fn round_with_frac_real(
+        &self,
+        ctx: &mut EvalContext,
+        row: &[Datum],
+    ) -> Result<Option<f64>> {
+        let x = try_opt!(self.children[0].eval_real(ctx, row));
+        let d = try_opt!(self.children[1].eval_int(ctx, row));
+
+        let base = 10.0_f64.powi(-d as i32);
+        let frac = x / base;
+        Ok(Some(frac.round() * base))
+    }
+
+    #[inline]
+    pub fn round_with_frac_dec<'a, 'b: 'a>(
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, Decimal>>> {
+        let x = try_opt!(self.children[0].eval_decimal(ctx, row));
+        let d = try_opt!(self.children[1].eval_int(ctx, row));
```

Comment:
Could you use some meaningful variable name? such as s/x/decimal and s/d/integer

Reply:
@DorianZheng Oh, sure I will replace with some other name. Currently I just follow arguments in MySQL document.

https://dev.mysql.com/doc/refman/5.7/en/mathematical-functions.html#function_round

>ROUND(X), ROUND(X,D)
>
>Rounds the argument X to D decimal places. The rounding algorithm depends on the data type of X. D defaults to 0 if not specified. D can be negative to cause D digits left of the decimal point of the value X to become zero.